src/dxvk/dxvk_fence.cpp

```diff
@@ -56,80 +57,94 @@ namespace dxvk {
         Logger::warn(str::format("Importing semaphores of type ", info.sharedType, " not supported by device"));
       }
     }
-
-    m_thread = dxvk::thread([this] { run(); });
   }
 
 
   DxvkFence::~DxvkFence() {
-    m_stop.store(true);
-    m_thread.join();
+    if (m_running.exchange(false, std::memory_order_acq_rel)) {
+      m_condVar.notify_one();
+      m_thread.join();
+    }
 
     m_vkd->vkDestroySemaphore(m_vkd->device(), m_semaphore, nullptr);
   }
 
 
   void DxvkFence::enqueueWait(uint64_t value, DxvkFenceEvent&& event) {
-    std::unique_lock<dxvk::mutex> lock(m_mutex);
+    if (value > getValue()) {
+      {
+        std::unique_lock<dxvk::mutex> lock(m_mutex);
+        m_events.emplace_back(value, std::move(event));
+      }
 
-    if (value > m_lastValue.load())
-      m_queue.emplace(value, std::move(event));
-    else
+      if (!m_running.exchange(true, std::memory_order_acq_rel)) {
+        m_thread = dxvk::thread([this] { run(); });
+      } else {
+        m_condVar.notify_one();
+      }
+    } else {
       event();
+    }
   }
-  
-
-  void DxvkFence::run() {
-    uint64_t value = 0ull;
 
+  void DxvkFence::wait(uint64_t value) {
     VkSemaphoreWaitInfo waitInfo = { VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO };
     waitInfo.semaphoreCount = 1;
     waitInfo.pSemaphores = &m_semaphore;
     waitInfo.pValues = &value;
+    VkResult vr = m_vkd->vkWaitSemaphores(
+      m_vkd->device(), &waitInfo, ~0ull);
+    if (vr != VK_SUCCESS) {
+      Logger::err(str::format("Failed to wait for semaphore: ", vr));
+    }
+  }
 
-    while (!m_stop.load()) {
-      std::unique_lock<dxvk::mutex> lock(m_mutex);
-
-      // Query actual semaphore value and start from there, so that
-      // we can skip over large increments in the semaphore value
-      VkResult vr = m_vkd->vkGetSemaphoreCounterValue(m_vkd->device(), m_semaphore, &value);
+  void DxvkFence::run() {
+    VkSemaphoreWaitInfo waitInfo = { VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO };
+    waitInfo.semaphoreCount = 1;
+    waitInfo.pSemaphores = &m_semaphore;
+    waitInfo.pValues = nullptr;
 
-      if (vr != VK_SUCCESS) {
-        Logger::err(str::format("Failed to query semaphore value: ", vr));
-        return;
-      }
+    while (m_running.load(std::memory_order_acquire)) {
```

Comment:
`acquire` doesn't make a huge amount of sense here, keeping the atomic is fine so we don't need to lock outside the loop but memory order literally doesn't matter for this one.